lib/src/mcu/auth/oauth2.rs

```diff
@@ -0,0 +1,434 @@
+use tokio::sync::Mutex;
+
+use crate::util::SensitiveString;
+
+use super::ApiClientAuth;
+
+#[derive(Clone, Debug)]
+pub struct AuthToken {
+    pub token: SensitiveString,
+    pub expires_at: chrono::DateTime<chrono::Utc>,
+}
+
+#[derive(serde::Deserialize)]
+enum TokenType {
+    Bearer,
+}
+
+#[derive(serde::Deserialize)]
+struct TokenResponse {
+    access_token: SensitiveString,
+    expires_in: i64,
+    #[allow(dead_code)]
+    token_type: TokenType,
+}
+
+struct Claims<'a> {
+    client_id: &'a str,
+    endpoint: &'a str,
+    issued_at: DateTime<Utc>,
+    expires_at: DateTime<Utc>,
+    token_id: &'a str,
+}
+
+impl<'a> serde::Serialize for Claims<'a> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let mut claims = serializer.serialize_struct("Claims", 6)?;
+        claims.serialize_field("iss", &self.client_id)?;
+        claims.serialize_field("aud", &self.endpoint)?;
+        claims.serialize_field("sub", &self.client_id)?;
+        claims.serialize_field("iat", &self.issued_at.timestamp())?;
+        claims.serialize_field("exp", &self.expires_at.timestamp())?;
+        claims.serialize_field("jti", &self.token_id)?;
+        claims.end()
+    }
+}
+
+pub struct OAuth2<'callback> {
+    endpoint: String,
+    client_id: String,
+    /// Private key (ES256)
+    client_key: SensitiveString,
+    token: Mutex<Option<AuthToken>>,
+    #[allow(clippy::type_complexity)]
+    token_callback: Box<dyn Fn(&AuthToken) + Send + Sync + 'callback>,
+}
+
+impl<'callback> OAuth2<'callback> {
+    #[must_use]
+    pub fn new(
+        endpoint: String,
+        client_id: String,
+        client_key: SensitiveString,
+        current_token: Option<AuthToken>,
+        token_callback: impl Fn(&AuthToken) + Send + Sync + 'callback,
+    ) -> Self {
+        Self {
+            endpoint,
+            client_id,
+            client_key,
+            token: Mutex::new(current_token),
+            token_callback: Box::new(token_callback),
+        }
+    }
+
+    fn generate_token_id() -> String {
+        let mut rng = rand::rngs::OsRng;
+        let bytes: [u8; 18] = rng.gen();
+        hex::encode(bytes)
+    }
+
+    async fn get_token(
+        endpoint: &str,
+        client_id: &str,
+        client_key: &jsonwebtoken::EncodingKey,
+    ) -> reqwest::Result<AuthToken> {
+        let issued_at = Utc::now();
+        let expires_at = issued_at + chrono::Duration::hours(1);
+        let token_id = Self::generate_token_id();
+        debug!("Generated token ID: {}", token_id);
+
+        let claims = jsonwebtoken::encode(
+            &Header::new(jsonwebtoken::Algorithm::ES256),
+            &Claims {
+                client_id,
+                endpoint,
+                issued_at,
+                expires_at,
+                token_id: &token_id,
+            },
+            client_key,
+        )
+        .unwrap();
+
+        let client = reqwest::Client::new();
```

Comment:
This new client doesn't respect the `--insecure` flag. Happy to open an issue so this can be fixed in a later PR if that's easier.

Reply:
I don't think we should be creating a new client here - probably should be injected.

Reply:
This should now work - the `reqwest::Client` is now injected, so we can use a correctly configured client.

Reply:
Can confirm it works now!